### ISSUE TITLE
Multiple Outputs

### ISSUE THREAD

**Issue by startupthekid** [BradLarson/GPUImage2#11]
I'm sure you have a ton on your plate with this rewrite but I thought I'd suggest a feature I'd love to see: one to many or many to one piping. That is, starting with a single video file, output several video files or starting with several video files, composite them into one video, similar to `AVComposition`.

**Comment by BradLarson** [comment]
One-to-many is already possible. You can hook up an arbitrary number of outputs to any given processing operation, so you could have multiple MovieOutputs simultaneously encoding. Can't guarantee that performance will be good, though.

For compositing, I won't be working on any of that, since I don't have an interest in AV Foundation movie processing nor the expertise to do that. Someone else will need to contribute that.

**Comment by startupthekid** [comment]
That's fair. I was thinking something along the lines of having `Movie` & `Audio` and they can be composited together internally using `AVMutableComposition`. Once the project is in a stable place, we can revisit this and I'll take a look at adding that functionality.

**Comment by BradLarson** [comment]
That'd be appreciated. Much of the movie code was contributed by others in the previous iteration, so I'm having a little difficulty bringing it across because I don't fully understand how some things were supposed to work. A lot of little patches and workarounds are strewn throughout, so I'm not sure what's critical and what's not.

I'll get it to a base level of functionality and see how it can be refined from there.

**Comment by startupthekid** [comment]
Yeah AVFoundation is tricky as all hell. There's a ton of different classes, even more so for OSX, and I'd say probably has the worst (or at least most confusing) documentation. And that's just dealing with video, CoreAudio is another beast entirely.

I was looking through `GPUImageMovie.h/m` and it seems a bit patch worked but the basics were definitely there. `AVAssetReader` and `AVAssetWriter` are probably your best bets for reading in sample buffers and passing them through the pipeline.  Compositions are...nasty. At that point it's less about editing and more about how things get composed together.

The one good thing about GPUImage though is that you don't have to deal with manipulating or filtering the sample buffers since everything is done using gpu shaders, fragments, what have you. Whenever you get a basic version of the movie code up, if you want to comment on this issue, I'd be glad to take a look. My gpu and opengl chops are a bit rusty but can't hurt to check it out.
